Objects/longobject.c

```diff
@@ -556,14 +575,14 @@ PyLong_AsLongAndOverflow(PyObject *vv, int *overflow)
         res = -1;
         i = _PyLong_DigitCount(v);
         sign = _PyLong_NonCompactSign(v);
-        x = 0;
+
+        x = _unroll_digits(v, &i);
         while (--i >= 0) {
-            prev = x;
-            x = (x << PyLong_SHIFT) | v->long_value.ob_digit[i];
-            if ((x >> PyLong_SHIFT) != prev) {
+            if (x > SIZE_MAX >> PyLong_SHIFT) {
```

Comment:
*x* type is unsigned long, not size_t.

```suggestion
            if (x > (ULONG_MAX >> PyLong_SHIFT)) {
```

Please check other functions such as PyLong_AsUnsignedLong().

Please add these parenthesis in other functions (using SIZE_MAX), for readability.